akka-typed/src/main/scala/akka/typed/cluster/ClusterSharding.scala

```diff
@@ -0,0 +1,196 @@
+ *           envelope.
+ * @tparam A The type of message accepted by the entity actor
+ */
+trait TypedMessageExtractor[E, A] {
+
+  /**
+   * Extract the entity id from an incoming `message`. If `null` is returned
+   * the message will be `unhandled`, i.e. posted as `Unhandled` messages on the event stream
+   */
+  def entityId(message: E): String
+
+  /**
+   * Extract the message to send to the entity from an incoming `message`.
+   * Note that the extracted message does not have to be the same as the incoming
+   * message to support wrapping in message envelope that is unwrapped before
+   * sending to the entity actor.
+   *
+   * If the returned value is `null`, and the entity isn't running yet the entity will be started
+   * but no message will be delivered to it.
+   */
+  def entityMessage(message: E): A
+
+  /**
+   * Extract the entity id from an incoming `message`. Only messages that passed the [[#entityId]]
+   * function will be used as input to this function.
+   */
+  def shardId(message: E): String
+}
+
+/**
+ * Java API:
+ *
+ * Default message extractor type, using envelopes to identify what entity a message is for
+ * and the hashcode of the entityId to decide which shard an entity belongs to.
+ *
+ * This is recommended since it does not force details about sharding into the entity protocol
+ *
+ * @tparam A The type of message accepted by the entity actor
+ */
+final class DefaultMessageExtractor[A](maxNumberOfShards: Int) extends TypedMessageExtractor[ShardingEnvelope[A], A] {
+  def entityId(envelope: ShardingEnvelope[A]) = envelope.entityId
+  def entityMessage(envelope: ShardingEnvelope[A]) = envelope.message
+  def shardId(envelope: ShardingEnvelope[A]) = (math.abs(envelope.entityId.hashCode) % maxNumberOfShards).toString
+}
+
+/**
+ * Java API:
+ *
+ * Default message extractor type, using a property of the message to identify what entity a message is for
+ * and the hashcode of the entityId to decide which shard an entity belongs to.
+ *
+ * This is recommended since it does not force details about sharding into the entity protocol
+ *
+ * @tparam A The type of message accepted by the entity actor
+ */
+abstract class DefaultNoEnvelopeMessageExtractor[A](maxNumberOfShards: Int) extends TypedMessageExtractor[A, A] {
+  def entityMessage(message: A) = message
+  def shardId(message: A) = {
+    val id = entityId(message)
+    if (id != null) (math.abs(id.hashCode) % maxNumberOfShards).toString
+    else null
+  }
+}
+
+/**
+ * A reference to an entityId and the local access to sharding, allows for actor-like interaction
+ *
+ * The entity ref must be resolved locally and cannot be sent to another node.
+ *
+ * TODO what about ask, should it actually implement ActorRef to be exactly like ActorRef and callers does not have
+ *      to know at all about it or is it good with a distinction but lookalike API?
```

Comment:
as I argued in the ticket it is not an ActorRef, because it doesn't have the same kind of lifecycle and semantics as an ActorRef

we probably have to support ask, but not watch